primitives/src/circuit/merkle_tree/mod.rs

```diff
@@ -75,10 +77,133 @@ where
 
     /// Enforce correct `merkle_proof` for the `elem` against
     /// `expected_merkle_root`.
-    fn enforce_merkle_proof(
+    fn enforce_membership_proof(
         &mut self,
         elem: Self::LeafVar,
         merkle_proof: Self::MerklePathVar,
         expected_merkle_root: Variable,
     ) -> Result<(), CircuitError>;
 }
+
+/// Gadget for the sparse merkle tree
+pub trait SparseMerkleTreeGadget<M>: MerkleTreeGadget<M>
+where
+    M: MerkleTreeScheme,
+{
+    /// checking non-membership proof
+    fn is_non_member(
+        &mut self,
+        elem: Self::LeafVar,
+        merkle_proof: Self::MerklePathVar,
+        merkle_root: Variable,
+    ) -> Result<BoolVar, CircuitError>;
+
+    /// Enforce correct `merkle_proof` for the empty leaf `elem` against
+    /// `expected_merkle_root`.
+    fn enforce_non_membership_proof(
+        &mut self,
+        empty_elem: Self::LeafVar,
+        merkle_proof: Self::MerklePathVar,
+        expected_merkle_root: Variable,
+    ) -> Result<(), CircuitError>;
+}
+
+pub(crate) trait MerkleTreeHelperGadget<M>
+where
+    M: MerkleTreeScheme,
+{
+    type MembershipProofBooleanEncoding;
+    type MembershipProofVar;
+    /// Produces an ordered list of variables based on the relative position of
+    /// a node and its siblings.
+    /// * `node` - node to be inserted in the final list.
+    /// * `sibling1` - first sibling
+    /// * `sibling2` - second sibling
+    /// * `node_is_left` - variable that is true if node is the leftmost one.
+    /// * `node_is_right` -  variable that is true if node is the rightmost one.
+    /// * `returns` - list of variables corresponding to the node and its
+    ///   siblings in the correct order.
+    fn permute(
+        &mut self,
+        node: Variable,
+        sib1: Variable,
+        sib2: Variable,
+        node_is_left: BoolVar,
+        node_is_right: BoolVar,
+    ) -> Result<[Variable; 3], CircuitError>;
+
+    /// Ensure that the position of each node of the path is correctly encoded
+    /// Used for testing purposes.
+    /// * `merkle_path` - list of node of an authentication path
+    /// * `pos` - position of the missing leaf
+    /// * `returns` - list of variables corresponding to the authentication path
+    fn constrain_membership_proof(
+        &mut self,
+        merkle_path: &Self::MembershipProofBooleanEncoding,
+        pos: M::Index,
+    ) -> Result<Self::MembershipProofVar, CircuitError>;
```

Comment:
curious why isn't this called `create_membership_proof_var`? (during creation, we check some structural validity, which is common). 

the first interface comes to mind when I saw `constrain_membership_proof` is something like `verify(proof_var)-> Result<(), E>`

Reply:
Hmm good question!
I suppose the reason is that when [making the interface more ergonomic](https://github.com/EspressoSystems/jellyfish/pull/158/), I mostly focused on the public API, whereas the MTGadgetHelper was an internal detail, really, its naming left untouched.
The name `create_membership_proof_variable` is already taken for the public API.
I'm open to suggestions for further naming options, though in this case I find `constrain_membership_proof` quite accurate.